File: backend/app/agent/date_normalizer.py
# backend/app/agent/date_normalizer.py
"""
日期格式规范化模块
确保前后端日期格式一致
"""
import re
from datetime import datetime
from typing import Any, Optional


# 支持的日期格式模式
DATE_FORMAT_PATTERNS = [
    # ISO 格式
    (r'^\d{4}-\d{2}-\d{2}$', '%Y-%m-%d'),  # YYYY-MM-DD
    (r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}', '%Y-%m-%dT%H:%M:%S'),  # ISO datetime
    # 中文格式
    (r'^\d{4}年\d{1,2}月\d{1,2}日$', '%Y年%m月%d日'),
    (r'^\d{4}/\d{1,2}/\d{1,2}$', '%Y/%m/%d'),  # YYYY/MM/DD
    (r'^\d{1,2}/\d{1,2}/\d{4}$', '%m/%d/%Y'),  # MM/DD/YYYY
    (r'^\d{4}\.\d{1,2}\.\d{1,2}$', '%Y.%m.%d'),  # YYYY.MM.DD
    # Excel 日期序列号（从1900-01-01开始的天数）
    (r'^\d+\.\d+$', None),  # 可能是 Excel 序列号
]


def normalize_date_value(value: Any) -> Any:
    """
    规范化日期值
    
    Args:
        value: 日期值（可能是字符串、数字、datetime对象等）
    
    Returns:
        规范化后的日期字符串（YYYY-MM-DD格式）或原值（如果不是日期）
    """
    if value is None:
        return value
    
    # 如果已经是标准格式的字符串，直接返回
    if isinstance(value, str):
        value_str = value.strip()
        looks_like_date = bool(re.search(r'\d', value_str)) and any(
            sep in value_str for sep in ('-', '/', '.', '年', '月', '日', 'T', ':')
        )
        # 纯数字字符串（如价格/库存/ID）不做日期推断，避免 7999 -> 7999-02-26
        if not looks_like_date:
            return value

        # 检查是否是标准 ISO 日期格式
        if re.match(r'^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2})?', value_str):
            # 提取日期部分（去掉时间部分）
            date_part = value_str.split('T')[0]
            return date_part
        
        # 尝试解析各种日期格式
        for pattern, date_format in DATE_FORMAT_PATTERNS:
            if date_format and re.match(pattern, value_str):
                try:
                    dt = datetime.strptime(value_str, date_format)
                    return dt.strftime('%Y-%m-%d')
                except ValueError:
                    continue
        
        # 尝试使用 Python 的 dateutil（如果可用）
        try:
            from dateutil import parser
            dt = parser.parse(value_str)
            return dt.strftime('%Y-%m-%d')
        except (ImportError, ValueError, TypeError):
            pass
    
    # 如果是数字，可能是 Excel 日期序列号
    elif isinstance(value, (int, float)):
        # Excel 日期序列号：从 1900-01-01 开始的天数
        # 但这里我们不确定，所以保持原值
        # 前端会处理 Excel 序列号
        return value
    
    # 如果是 datetime 对象
    elif isinstance(value, datetime):
        return value.strftime('%Y-%m-%d')
    
    # 其他类型，返回原值
    return value


def normalize_date_range(value: Any) -> Optional[dict]:
    """
    规范化日期范围
    
    Args:
        value: 日期范围值（可能是字符串、dict等）
    
    Returns:
        规范化后的日期范围字典，格式：{"start": "YYYY-MM-DD", "end": "YYYY-MM-DD"}
    """
    if value is None:
        return None
    
    if isinstance(value, dict):
        normalized = {}
        if 'start' in value:
            normalized['start'] = normalize_date_value(value['start'])
        if 'end' in value:
            normalized['end'] = normalize_date_value(value['end'])
        if 'min' in value:
            normalized['start'] = normalize_date_value(value['min'])
        if 'max' in value:
            normalized['end'] = normalize_date_value(value['max'])
        return normalized if normalized else None
    
    if isinstance(value, str):
        # 尝试解析日期范围字符串（如 "2024-01-01 to 2024-12-31"）
        range_match = re.match(r'(.+?)\s+(?:to|-|~)\s+(.+)$', value.strip())
        if range_match:
            start_str = range_match.group(1).strip()
            end_str = range_match.group(2).strip()
            return {
                'start': normalize_date_value(start_str),
                'end': normalize_date_value(end_str)
            }
    
    return None


def normalize_validation_params(validation_type: str, validation_params: dict) -> dict:
    """
    规范化数据验证参数中的日期
    
    Args:
        validation_type: 验证类型（如 "date"）
        validation_params: 验证参数字典
    
    Returns:
        规范化后的验证参数字典
    """
    if validation_type != 'date':
        return validation_params
    
    normalized = validation_params.copy()
    
    # 处理日期范围
    if 'min' in normalized:
        normalized['min'] = normalize_date_value(normalized['min'])
    if 'max' in normalized:
        normalized['max'] = normalize_date_value(normalized['max'])
    if 'start' in normalized:
        normalized['start'] = normalize_date_value(normalized['start'])
    if 'end' in normalized:
        normalized['end'] = normalize_date_value(normalized['end'])
    
    # 处理日期范围对象
    if 'range' in normalized:
        normalized['range'] = normalize_date_range(normalized['range'])
    
    return normalized

File: frontend/src/utils/dateNormalizer.js
// frontend/src/utils/dateNormalizer.js
/**
 * 日期格式规范化模块
 * 确保前后端日期格式一致
 */

/**
 * 规范化日期值
 * 将各种日期格式统一转换为 YYYY-MM-DD 格式
 */
export function normalizeDateValue(value) {
  if (value === null || value === undefined) {
    return value
  }
  
  // 如果已经是标准格式的字符串，直接返回
  if (typeof value === 'string') {
    const trimmed = value.trim()
    const looksLikeDate =
      /\d/.test(trimmed) &&
      (trimmed.includes('-') ||
        trimmed.includes('/') ||
        trimmed.includes('.') ||
        trimmed.includes('年') ||
        trimmed.includes('月') ||
        trimmed.includes('日') ||
        trimmed.includes('T') ||
        trimmed.includes(':'))

    // 纯数字字符串（如价格/库存/ID）不做日期推断，避免 7999 -> 7999-02-26
    if (!looksLikeDate) {
      return value
    }

    // 检查是否是标准 ISO 日期格式 (YYYY-MM-DD)
    if (/^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2})?/.test(trimmed)) {
      // 年份 < 1930 的 ISO 字符串几乎必然是"数值被误转为 Excel 序列号日期"的残留产物
      // （如数量 2 → "1900-01-02"，价格 500 → "1901-05-15"），不做日期规范化，原样返回
      const isoYear = parseInt(trimmed.slice(0, 4), 10)
      if (isoYear < 1930) return value
      return trimmed.split('T')[0]
    }
    
    // 尝试解析各种日期格式
    // 格式：YYYY/MM/DD, MM/DD/YYYY, YYYY.MM.DD, YYYY年MM月DD日等
    const dateFormats = [
      /^(\d{4})\/(\d{1,2})\/(\d{1,2})$/,  // YYYY/MM/DD
      /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/,  // MM/DD/YYYY
      /^(\d{4})\.(\d{1,2})\.(\d{1,2})$/,  // YYYY.MM.DD
      /^(\d{4})年(\d{1,2})月(\d{1,2})日$/, // YYYY年MM月DD日
    ]
    
    for (const pattern of dateFormats) {
      const match = trimmed.match(pattern)
      if (match) {
        let year, month, day
        if (pattern === dateFormats[0]) {
          // YYYY/MM/DD
          year = match[1]
          month = match[2].padStart(2, '0')
          day = match[3].padStart(2, '0')
        } else if (pattern === dateFormats[1]) {
          // MM/DD/YYYY
          month = match[1].padStart(2, '0')
          day = match[2].padStart(2, '0')
          year = match[3]
        } else if (pattern === dateFormats[2]) {
          // YYYY.MM.DD
          year = match[1]
          month = match[2].padStart(2, '0')
          day = match[3].padStart(2, '0')
        } else if (pattern === dateFormats[3]) {
          // YYYY年MM月DD日
          year = match[1]
          month = match[2].padStart(2, '0')
          day = match[3].padStart(2, '0')
        }
        
        // 验证日期有效性
        const date = new Date(`${year}-${month}-${day}`)
        if (!isNaN(date.getTime())) {
          return `${year}-${month}-${day}`
        }
      }
    }
    
    // 禁止宽松 Date 兜底：
    // 像 "7.5"、"10/2" 这类业务数字在不同运行时会被 Date 误解析为日期，
    // 造成占比/金额被写成 2001-07-04 等错误值。
    // 仅对显式日期格式做规范化，其余字符串保持原样。
    return value
  }
  
  // 如果是数字，可能是 Excel 日期序列号（从1900-01-01开始的天数）
  if (typeof value === 'number') {
    // Excel 日期序列号：1 = 1900-01-01
    // 但这里我们不确定，所以保持原值
    // 如果确实是日期序列号，前端会在显示时转换
    return value
  }
  
  // 如果是 Date 对象
  if (value instanceof Date) {
    return value.toISOString().split('T')[0]
  }
  
  // 其他类型，返回原值
  return value
}

/**
 * 规范化日期范围
 */
export function normalizeDateRange(value) {
  if (value === null || value === undefined) {
    return null
  }
  
  if (typeof value === 'object' && !Array.isArray(value)) {
    const normalized = {}
    if (value.start !== undefined) {
      normalized.start = normalizeDateValue(value.start)
    }
    if (value.end !== undefined) {
      normalized.end = normalizeDateValue(value.end)
    }
    if (value.min !== undefined) {
      normalized.start = normalizeDateValue(value.min)
    }
    if (value.max !== undefined) {
      normalized.end = normalizeDateValue(value.max)
    }
    return Object.keys(normalized).length > 0 ? normalized : null
  }
  
  if (typeof value === 'string') {
    // 尝试解析日期范围字符串（如 "2024-01-01 to 2024-12-31"）
    const rangeMatch = value.match(/(.+?)\s+(?:to|-|~)\s+(.+)$/)
    if (rangeMatch) {
      return {
        start: normalizeDateValue(rangeMatch[1].trim()),
        end: normalizeDateValue(rangeMatch[2].trim())
      }
    }
  }
  
  return null
}

/**
 * 规范化数据验证参数中的日期
 */
export function normalizeValidationParams(validationType, validationParams) {
  if (validationType !== 'date') {
    return validationParams
  }
  
  if (!validationParams || typeof validationParams !== 'object') {
    return validationParams
  }
  
  const normalized = { ...validationParams }
  
  // 处理日期范围
  if (normalized.min !== undefined) {
    normalized.min = normalizeDateValue(normalized.min)
  }
  if (normalized.max !== undefined) {
    normalized.max = normalizeDateValue(normalized.max)
  }
  if (normalized.start !== undefined) {
    normalized.start = normalizeDateValue(normalized.start)
  }
  if (normalized.end !== undefined) {
    normalized.end = normalizeDateValue(normalized.end)
  }
  
  // 处理日期范围对象
  if (normalized.range) {
    normalized.range = normalizeDateRange(normalized.range)
  }
  
  return normalized
}
